torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -933,3 +968,438 @@ def get_rule(
+                # the type promotion rule should not suggest promoting this arg.
+                graph = node.graph
+                with graph.inserting_before(node):
+                    target = getattr(torch.ops.aten.tensor, type(fx_arg).__name__)
+                    new_fx_arg = self._create_node(
+                        graph, "call_function", target, (fx_arg,), {"dtype": dtype}
+                    )
+                    diagnostic.with_additional_message(
+                        f"Argument {fx_arg}(Scalar of equivalent dtype: {equivalent_dtype}) "
+                        f"is promoted to {dtype}."
+                    )
+                    return new_fx_arg
+            diagnostic.with_additional_message(
+                f"Argument {fx_arg} is not promoted. Already {dtype}."
+            )
+            return fx_arg
+        elif isinstance(fx_arg, (tuple, list)):
+            new_fx_arg = []
+            diagnostic.with_additional_message(
+                f"Argument {fx_arg} is a tuple/list. Promoting each element."
+            )
+            for fx_arg_elem in fx_arg:
+                new_fx_arg.append(
+                    self._promote_arg(diagnostic, node, fx_arg_elem, dtype)
+                )
+            return type(fx_arg)(new_fx_arg)
+        else:
+            raise NotImplementedError(f"Unknown fx arg type: {type(fx_arg)}")
+
+    @_beartype.beartype
+    def _promote_node(
+        self,
+        diagnostic: diagnostics.Diagnostic,
+        node: torch.fx.Node,
+        rule: TypePromotionRule,
+    ) -> torch.fx.Node:
+        """Promote node inputs and outputs according to type promotion rule."""
+        args, kwargs = self.fetch_args_kwargs_from_env(node)
+        type_promotion_info = rule.preview_type_promotion(args, kwargs)
+        new_args = []
+        new_kwargs = {}
+        for i, arg in enumerate(node.args):
+            new_args.append(
+                self._promote_arg(
+                    diagnostic, node, arg, type_promotion_info.args_dtypes.get(i, None)
+                )
+            )
+
+        for name, arg in node.kwargs.items():
+            new_kwargs[name] = self._promote_arg(
+                diagnostic, node, arg, type_promotion_info.kwargs_dtypes.get(name, None)
+            )
+        new_args = tuple(new_args)
+
+        if node.args != new_args or node.kwargs != new_kwargs:
+            diagnostic.message = f"Applied type promotion for {node}. "
+            node.args = new_args
+            node.kwargs = new_kwargs
+            self._rerun_node_after_type_promotion(
+                diagnostic, node, type_promotion_info.out_dtype
+            )
+        else:
+            diagnostic.message = f"Type promotion not needed for {node}. "
+
+        return node
+
+    @diagnostics.diagnose_call(
+        rule=diagnostics.rules.fx_node_insert_type_promotion,
+        level=diagnostics.levels.NONE,
+    )
+    def run_node(self, node: torch.fx.Node) -> Any:
```

Comment:
It'd be helpful to explain the high-level calling logic in English here I think: Is it `run_node -> _promote_node -> _promote_args -> _create_node -> _run_node_and_update_meta_val -> rerun_node`? When we do rerun, does it happen exactly once for each node?

Reply:
Yea except for swapping this order here `rerun_node -> _run_node_and_update_meta_val`. Each node is guaranteed to run at least once. We can change the code to run every node exactly once, but it adds more complexity and I don't feel it is necessary since this is always running in fake mode.

I'll update the docstring with high level logic of things to do.

Reply:
> Each node is guaranteed to run at least once.

Sounds good! I would just make this clear in the docs